### PR TITLE
Links ramp rate

### PR DESCRIPTION
**Pull request by smdumlao**
## Changes proposed in this Pull Request
Added ramp rate limits for links. The functionality is limited to non-committable links. 

As suggested, I reused the code for the Generators. I modified `define_ramp_limit_constraints` to include links as valid `component` input. However, I am not sure how to incorporate the committable functionality at the moment since it uses several variables; thus, a flag for `committable` was added to keep this functionality in Generators but not in Links. 

Kindly advised for necessary tests. I used it for my own purpose (representing hydrogen power plant using store and links) and it was able to limit the ramp rate for both up and down. 

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [ ] Unit tests for new features were added (if applicable).
- [ ] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Comment by smdumlao** [comment]
I am currently using the code below to use the `extra_functionality` of `lopf` and `override_component_attrs` in setting up the network. The pull request is equivalent to these implementation. I was wondering if I am using it correctly?

```
import pandas as pd 

from tempfile import mkstemp
from pypsa.linopt import linexpr, get_var, define_constraints
from pypsa.linopf import define_nominal_for_extendable_variables, define_fixed_variable_constraints
from pypsa.descriptors import get_non_extendable_i, get_extendable_i, get_active_assets, get_activity_mask, Dict
from pypsa.components import component_attrs

# add ramp limit into the component attrs
mask = component_attrs['Generator'].index.isin(['ramp_limit_up', 'ramp_limit_down'])
component_attrs['Generator'][mask]
component_attrs['Link'] = pd.concat([component_attrs['Link'], component_attrs['Generator'][mask]])


def define_link_ramp_limit_constraints(n, sns):
    """
    Defines ramp limits for links with valid ramplimit;
    This is replicated from the define_ramp_limit_constraints() for generators.
    Current implementation does not consider the committable links yet; thus, it merely focuses 
    on limiting the ramp rate per time period.

    """
    c = 'Link'
    rup_i = n.df(c).query('ramp_limit_up == ramp_limit_up').index
    rdown_i = n.df(c).query('ramp_limit_down == ramp_limit_down').index
    if rup_i.empty & rdown_i.empty:
        return
    fix_i = get_non_extendable_i(n, c)
    ext_i = get_extendable_i(n, c)
    p = get_var(n, c, 'p').loc[sns[1:]]
    p_prev = get_var(n, c, 'p').shift(1).loc[sns[1:]]
    active = get_activity_mask(n, c, sns[1:])

    # fix up
    gens_i = rup_i.intersection(fix_i)
    if not gens_i.empty:
        lhs = linexpr((1, p[gens_i]), (-1, p_prev[gens_i]))
        rhs = n.df(c).loc[gens_i].eval('ramp_limit_up * p_nom')
        kwargs = dict(spec='nonext.', mask=active[gens_i])
        define_constraints(n, lhs, '<=', rhs,  c, 'mu_ramp_limit_up', **kwargs)

    # ext up
    gens_i = rup_i.intersection(ext_i)
    if not gens_i.empty:
        limit_pu = n.df(c)['ramp_limit_up'][gens_i]
        p_nom = get_var(n, c, 'p_nom')[gens_i]
        lhs = linexpr((1, p[gens_i]), (-1, p_prev[gens_i]), (-limit_pu, p_nom))
        kwargs = dict(spec='ext.', mask=active[gens_i])
        define_constraints(n, lhs, '<=', 0, c, 'mu_ramp_limit_up', **kwargs)

    # fix down
    gens_i = rdown_i.intersection(fix_i)
    if not gens_i.empty:
        lhs = linexpr((1, p[gens_i]), (-1, p_prev[gens_i]))
        rhs = n.df(c).loc[gens_i].eval('-1 * ramp_limit_down * p_nom')
        kwargs = dict(spec='nonext.', mask=active[gens_i])
        define_constraints(n, lhs, '>=', rhs, c, 'mu_ramp_limit_down', **kwargs)

    # ext down
    gens_i = rdown_i.intersection(ext_i)
    if not gens_i.empty:
        limit_pu = n.df(c)['ramp_limit_down'][gens_i]
        p_nom = get_var(n, c, 'p_nom')[gens_i]
        lhs = linexpr((1, p[gens_i]), (-1, p_prev[gens_i]), (limit_pu, p_nom))
        kwargs = dict(spec='ext.', mask=active[gens_i])
        define_constraints(n, lhs, '>=', 0, c, 'mu_ramp_limit_down', **kwargs)
```

The code below is used to setup the `network` using the `override_component_attrs`.
```
network = pypsa.Network(override_component_attrs=component_attrs)
```

The code below is used to run the `lopf`.
```
network.lopf(network.snapshots, pyomo=False, solver_name='gurobi',
                     extra_functionality = define_link_ramp_limit_constraints)
```

File: pypsa/linopf.py [hunk 1]
```diff
@@ -233,19 +233,20 @@ def define_committable_generator_constraints(n, sns):
 
 
 
-def define_ramp_limit_constraints(n, sns):
+def define_ramp_limit_constraints(n, sns, c='Generator', commitable=True):
```
Comment (FabianHofmann) [inline pypsa/linopf.py:236]:
```suggestion
def define_ramp_limit_constraints(n, sns, c):
```

File: pypsa/linopf.py [hunk 2]
```diff
@@ -233,19 +233,20 @@ def define_committable_generator_constraints(n, sns):
...
     """
-    Defines ramp limits for generators with valid ramplimit
+    Defines ramp limits for generators and links with valid ramplimit.
```
Comment (FabianHofmann) [inline pypsa/linopf.py:238]:
```suggestion
    Defines ramp limits for a given component with valid ramplimit.
```

File: pypsa/linopf.py [hunk 3]
```diff
@@ -233,19 +233,20 @@ def define_committable_generator_constraints(n, sns):
...
 
     """
-    c = 'Generator'
+    test_components = ['Generator', 'Link']
+    assert c in test_components, 'Ramp limit constraints were only tested for Generator and Link.'
+
```
Comment (FabianHofmann) [inline pypsa/linopf.py:243]:
```suggestion
```

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
Hey @smdumlao, thanks for the contributions and sorry this takes so long!

Looks good, however I would like to make the changes a bit more leaner/efficient.
Could you change the following:

* make `c` an positional argument in `define_ramp_limit_constraints` (see proposed change)
* replace the line where there used to be
   ```
   com_i = n.df(c).query('committable').index.difference(ext_i)
   ```
   with 
   ```
    if "committable" in n.df(c):
        com_i = n.df(c).query('committable').index.difference(ext_i)
    else:
        com_i = []
    ```
* undo the `if committable` case differentiation
* apply the other proposed changes

The idea is that if a component does not have the `committable` column, then `com_i` is empty and the ramp constraint for committable assets is skipped. You agree with that?

**Comment by smdumlao** [comment]
Hi @fneum, thank you for checking my contribution. I have made your suggested changes. 

> The idea is that if a component does not have the `committable` column, then `com_i` is empty and the ramp constraint for committable assets is skipped. You agree with that?

>   ```
>    if "committable" in n.df(c):
>        com_i = n.df(c).query('committable').index.difference(ext_i)
>    else:
>        com_i = []
>   ```

I agree with this approach. The `if "committable" in n.df('link')` will always return a `False` thus giving an empty `com_i` and subsequently, the `gens_i` in the succeeding lines with be empty.

File: pypsa/linopf.py [hunk 4]
```diff
@@ -297,6 +282,24 @@ def define_ramp_limit_constraints(n, sns):
         kwargs = dict(spec='ext.', mask=active[gens_i])
         define_constraints(n, lhs, '>=', 0, c, 'mu_ramp_limit_down', **kwargs)
 
+    if "committable" in n.df(c):
+        com_i = n.df(c).query('committable').index.difference(ext_i)
+    else:
+        com_i = []
```
Comment (FabianHofmann) [inline pypsa/linopf.py:288]: Could you move these lines up to where `com_i` was defined initially?

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
Almost perfect!

One comment thing regarding the code (see below). 
Could you add one comment to the release notes? Something like:

The `Link` component has a two new attributes `ramp_limit_up` and `ramp_limit_down` which allows to set the maximum/minimum power increase, equivalently to the implementation for generators. The new attributes are only considered when performing the `lopf` with `pyomo=False`.

**Comment by smdumlao** [comment]
@FabianHofmann I added a bullet in the release notes under "Upcoming Release". I hope this is the proper file.

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
Thanks you @smdumlao, I would still have two picky things and one important:

1. Could you restore the original order of the constraint definition in the function? `fix up`, `ext up`, `com up`. The `com up` snippet moved downwards (see under `Files changed`)  
2. Could you add a warning in the `opf.py` here https://github.com/PyPSA/PyPSA/blob/5a1f9d643971e5d2c1e1169b92c0eb17ec45fd74/pypsa/opf.py#L1657 that the ramp limits for links are ignored. Something like 
``` 
if network.links[['ramp_limit_up', 'ramp_limit_down']].notnull().any().any():
    logger.warning("Encountered nonzero ramp limits for links. These are ignored running the optimization with `pyomo=False`.")
```
3. There was a line deleted (902) `    define_committable_generator_constraints(n, snapshots)
`. Can you reinsert that one please?

Then we are done :)

**Review by FabianHofmann — APPROVED** [review]
Great! Thank you for the contribution @smdumlao. I'll merge after the tests pass

**Comment by smdumlao** [comment]
Thank you for the code review @FabianHofmann! I am glad to have contributed something to this project.

**Comment by FabianHofmann** [comment]
Looking forward to the next time :)

**Pull request merged**